p2p/pre2p/raintree/network.go

```diff
@@ -0,0 +1,207 @@
+package raintree
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"log"
+	"math/rand"
+	"time"
+
+	"github.com/pokt-network/pocket/shared/config"
+
+	typesPre2P "github.com/pokt-network/pocket/p2p/pre2p/types"
+	cryptoPocket "github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/types"
+
+	"google.golang.org/protobuf/proto"
+)
+
+var _ typesPre2P.Network = &rainTreeNetwork{}
+
+type rainTreeNetwork struct {
+	selfAddr cryptoPocket.Address
+	addrBook typesPre2P.AddrBook
+
+	// TECHDEBT(olshansky): Consider optimizing these away if possible.
+	// Helpers / abstractions around `addrBook` for simpler implementation through additional
+	// storage & pre-computation.
+	addrBookMap  typesPre2P.AddrBookMap
+	addrList     []string
+	maxNumLevels uint32
+
+	// TECHDEBT(drewsky): What should we use for de-duping messages within P2P?
+	mempool types.Mempool
+}
+
+func NewRainTreeNetwork(addr cryptoPocket.Address, addrBook typesPre2P.AddrBook, config *config.Config) typesPre2P.Network {
+	n := &rainTreeNetwork{
+		selfAddr: addr,
+		addrBook: addrBook,
+		// This subset of fields are initialized by `handleAddrBookUpdates` below
+		addrBookMap:  make(typesPre2P.AddrBookMap),
+		addrList:     make([]string, 0),
+		maxNumLevels: 0,
+		// TODO(team): Mempool size should be configurable
+		mempool: types.NewMempool(1000000, 1000),
+	}
+
+	if err := n.handleAddrBookUpdates(); err != nil {
+		// DISCUSS(drewsky): if this errors, the node could still function but not participate in
+		// message propagation. Should we return an error or just log?
+		log.Println("[ERROR] Error initializing rainTreeNetwork: ", err)
+	}
+
+	return typesPre2P.Network(n)
+}
+
+func (n *rainTreeNetwork) NetworkBroadcast(data []byte) error {
+	return n.networkBroadcastInternal(data, n.maxNumLevels, getNonce())
```

Comment:
I think this would be better named as `broadcastAtLevel`

Reply:
DONE + AGGREGATE